src/bigquery/src/query.rs

```diff
@@ -0,0 +1,178 @@
+    pub(crate) job_service: Arc<JobService>,
+    pub(crate) project_id: String,
+    pub(crate) job_id: String,
+    pub(crate) location: String,
+    pub(crate) query_id: Option<String>,
+    pub(crate) page_token: String,
+    pub(crate) schema: Option<Schema>,
+    pub(crate) total_rows: u64,
+    pub(crate) completed: bool,
+    pub(crate) cached_rows: VecDeque<wkt::Struct>,
+    pub(crate) num_dml_affected_rows: i64,
+}
+
+/// Represents errors that can occur when running queries.
+#[derive(thiserror::Error, Debug)]
+#[non_exhaustive]
+pub enum QueryError {
+    /// Only complete Query Jobs can be read.
+    #[error("Query is not complete: Only complete Query Jobs can be read.")]
+    NotComplete,
+}
+
+impl Query {
+    pub(crate) fn new(job_service: Arc<JobService>, res: model::QueryResponse) -> Self {
+        let mut query = Self {
+            job_service,
+            project_id: String::default(),
+            job_id: String::default(),
+            location: String::default(),
+            page_token: String::default(),
+            query_id: None,
+            schema: None,
+            total_rows: 0,
+            completed: false,
+            cached_rows: VecDeque::new(),
+            num_dml_affected_rows: 0,
+        };
+        query.consume_query_response(from_get_query_response(res.clone()));
+        if !res.query_id.is_empty() {
+            query.query_id = Some(res.query_id);
+        }
+        query
+    }
+
+    pub fn job_reference(&self) -> Option<model::JobReference> {
+        if self.job_id.is_empty() {
+            return None;
+        }
+        Some(
+            model::JobReference::new()
+                .set_project_id(self.project_id.clone())
+                .set_job_id(self.job_id.clone())
+                .set_location(self.location.clone()),
+        )
+    }
+
+    pub fn query_id(&self) -> Option<&str> {
+        self.query_id.as_deref()
+    }
+
+    pub fn schema(&self) -> Option<model::TableSchema> {
+        self.schema.clone().map(|s| s.schema)
+    }
+
+    pub fn total_rows(&self) -> u64 {
+        self.total_rows
+    }
+
+    pub fn completed(&self) -> bool {
+        self.completed
+    }
+
+    pub fn num_dml_affected_rows(&self) -> i64 {
+        self.num_dml_affected_rows
+    }
+
+    pub async fn wait(&mut self) -> Result<()> {
+        if self.completed {
+            return Ok(());
+        }
+        let sleep = async |d| tokio::time::sleep(d).await;
+        loop {
+            let complete = self.poll_job().await?;
+            if complete {
+                return Ok(());
+            }
+            // TODO: exponential backoff
+            sleep(Duration::from_millis(300)).await;
+        }
+    }
+
+    pub async fn read(self) -> Result<RowIterator> {
+        if !self.completed {
+            return Err(crate::Error::ser(QueryError::NotComplete));
+        }
+        RowIterator::try_new(Arc::new(self)).await
+    }
+
+    async fn poll_job(&mut self) -> Result<bool> {
```

Comment:
Would folks want a timeout for poll_job / wait?

Reply:
yeah, in Go we added support for context cancelation and in Node with Promise/AbortController. Need to check what to do in Rust, but that is really language specific.